### ISSUE TITLE
Unable to train Dispnet and Posenet.

### ISSUE THREAD

**Issue by ezorfa** [anuragranj/cc#19]
Hi @anuragranj !

After freshly checking out your code, I was trying to train dispnet and posenet with the followng command:
`python3 train.py /cdtemp/ezorfa/gateway30/datasets/kitti/kitti_cc --
dispnet DispResNet6 --posenet PoseNetB6   --masknet MaskNet6 --flownet Back2Future  -b 4 -pc 1.0 -pf 0.0 -c 0.0 -s 0.1  --log-output -f 50 --nlevels 6 --lr 1e-4 -wssim 0.997    --epochs 100 --smoothness-type edgeaware  --fix-masknet --fix-flownet   --log-terminal --name cc_depth  --epoch-size 1000`

But even after 30K iterations there seem to be no hint of training: 

<img width="515" alt="Screenshot 2019-12-09 at 08 19 06" src="https://user-images.githubusercontent.com/50787308/70415304-9c4fa480-1a5c-11ea-8a5b-b0541b3b9915.png">

<img width="515" alt="Screenshot 2019-12-09 at 08 18 50" src="https://user-images.githubusercontent.com/50787308/70415305-9c4fa480-1a5c-11ea-8060-8460b38dc5f3.png">

Previously, I also trained for over 200K, but I dint see any result or any hint of training. It just stays white image  all the time. **Could you suggest me or hint at what I might be doing wrong??**

Appreciate your help,
Thankyou!

**Comment by ezorfa** [comment]
@anuragranj Do I need to play with smoothness parameter initially or may be seeding is the problem?

**Comment by anuragranj** [comment]
There should be some weight on mask when training for the first time. This will force the loss to use depth and camera motion networks. So, don't fix masknet when training for the first time. Could you try this and see how it looks

```sh
python3 train.py <DATA_PATH> --sequence-length 5 --dispnet DispResNet6 --posenet PoseNetB6 --masknet MaskNet6 --flownet Back2Future -b8 -m0.6 -pf 0.0 -pc 1.0 -s0.005 --epoch-size 1000 --log-output -f 0 --nlevels 6 --lr 1e-4 -wssim 0.997 --with-flow-gt --with-depth-gt --epochs 100 --fix-flownet --smoothness-type edgeaware --name <NAME>
```

**Comment by ezorfa** [comment]
Ohh IC ! Thankyou @anuragranj . I will come back to you.

However, I dont still see the need to activate masknet here. Is there a way that I get away from using masknet?

**Comment by ezorfa** [comment]
@anuragranj  Like, Can't I do:

`loss = w1*loss_1 +  w3*loss_3 ; where loss_3 = smooth_loss(depth)`

I believe this^ loss function is in accordance with Sfmlearner.

**Comment by anuragranj** [comment]
> Ohh IC ! Thankyou @anuragranj . I will come back to you.
> 
> However, I dont still see the need to activate masknet here. Is there a way that I get away from using masknet?

Yes. It's a little tricky. You would need to pass all ones in place of explainability mask in the call of the following function.

https://github.com/anuragranj/cc/blob/master/train.py#L490-L491

This is because depth loss is computed only on the masked pixels.

**Comment by ezorfa** [comment]
@anuragranj  Hi !

I tried with the command you specified. It worked! I started to see some results right from 2500 steps.

However, I have few questions. Please answer them briefly:

[1] ` loss = w1*loss_1 + w3*loss_3 ; where loss_3 = smooth_loss(depth); where loss_1 = loss_camera(tgt_img_var, ref_imgs_var, intrinsics_var, intrinsics_inv_var, depth,`  **None**` , pose, lambda_oob=args.lambda_oob, qch=args.qch, wssim=args.wssim)` 

Dont you think this should work?

[2] Why should the `epoch_size = 1000` ? Should'nt it be the full size of the dataset?

[3] Initially ( to stop training for depth and pose), should the stop criteria be minimum photometric camera error?

Thankyou!

**Comment by ezorfa** [comment]
@anuragranj  sorry for too many questions, regarding point [2], I believe this is to get the networks trained a little initially.

**Comment by anuragranj** [comment]
> @anuragranj Hi !
> 
> I tried with the command you specified. It worked! I started to see some results right from 2500 steps.
> 
> However, I have few questions. Please answer them briefly:
> 
> [1] ` loss = w1*loss_1 + w3*loss_3 ; where loss_3 = smooth_loss(depth); where loss_1 = loss_camera(tgt_img_var, ref_imgs_var, intrinsics_var, intrinsics_inv_var, depth,` **None**` , pose, lambda_oob=args.lambda_oob, qch=args.qch, wssim=args.wssim)`
> 
> Dont you think this should work?
> 
This should work, yes. I dint realize that exp_mask=None was accounted for in the loss_functions.py.

> [2] Why should the `epoch_size = 1000` ? Should'nt it be the full size of the dataset?
> 
Just to make validation step faster. The dataloader is randomly shuffled anyway.

> [3] Initially ( to stop training for depth and pose), should the stop criteria be minimum photometric camera error?
> 
I used validation loss on depth ground truths.

> Thankyou!
